src/basic/bpf-program.c

```diff
@@ -117,19 +117,24 @@ int bpf_program_cgroup_attach(BPFProgram *p, int type, const char *path, uint32_
         return 0;
 }
 
-int bpf_program_cgroup_detach(int type, const char *path) {
+int bpf_program_cgroup_detach(BPFProgram *p, int type, const char *path) {
         _cleanup_close_ int fd = -1;
         union bpf_attr attr;
 
+        assert(type >= 0);
         assert(path);
 
+        /* Note that 'p' may be NULL, in which case any program is detached. However, note that if BPF_F_ALLOW_MULTI is
+         * used 'p' is not optional. */
+
         fd = open(path, O_DIRECTORY|O_RDONLY|O_CLOEXEC);
         if (fd < 0)
                 return -errno;
 
         attr = (union bpf_attr) {
                 .attach_type = type,
                 .target_fd = fd,
+                .attach_bpf_fd = p ? p->kernel_fd : -1,
```

Comment:
Buuuuuuut, that's not backwards compatible. What sort of API is that?

Reply:
the code on the kernel side is beautiful about this actually, it actively acknowledges that attach_bpf_fd might be positive but invalid, and hence won't use/check the field unless ALLOW_MULTI is set for compat reason. Such a young API, and already so fucked.